File: top100/259-3sum-smaller.java
public class Solution {
    public int threeSumSmaller(int[] nums, int target) {
        int n = nums.length;
        if (n < 3) return 0;
        //sort the array
        Arrays.sort(nums);
        
        int count = 0;        
        for (int i = 0; i<n-2; i++){
            int low = i+1;
            int high = n-1;
            while(low<high){
                if (nums[i]+nums[low]+nums[high] < target){
                    count = count + high - low;
                    low++;
                }
                else {
                    high--;
                }
            }
        }
        
        return count;
    }
}

File: leetcode/259. 3Sum Smaller.py
class Solution(object):
    def threeSumSmaller(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        if not nums or len(nums) <= 2: 
            return 0
        
        sNums = sorted(nums)
        
        n = len(nums)
        c = 0
        
        for i, num in enumerate(sNums):
            subTarget = target - num
            
            l = i+1
            r = n - 1
            
            while (l < r):
                if sNums[l] + sNums[r] < subTarget:
                    c += (r-l)
                    l += 1
                else:
                    r -= 1

        return c
